### DISCUSSION TITLE
Middleware ⚡️ Node.js с примерами кода

### DISCUSSION THREAD

**Discussion by giscus** [xsltdev/nodejsdev.ru#54] [Announcements]
# Middleware ⚡️ Node.js с примерами кода

Справочник Node.js с примерами кода

https://nodejsdev.ru/guides/webdraftt/middleware/

**Comment by GenGPT** [comment]
//sql

CREATE TABLE rooms (

  id SERIAL PRIMARY KEY,

  number VARCHAR(50) NOT NULL,

  type VARCHAR(50) NOT NULL,

  price DECIMAL NOT NULL

);


CREATE TABLE bookings (

  id SERIAL PRIMARY KEY,

  room_id INT REFERENCES rooms(id),

  guest_name VARCHAR(100) NOT NULL,

  check_in_date DATE NOT NULL,

  check_out_date DATE NOT NULL

);


CREATE TABLE users (

  id SERIAL PRIMARY KEY,

  username VARCHAR(50) NOT NULL UNIQUE,

  password VARCHAR(255) NOT NULL

);


INSERT INTO rooms (id, number, type, price) VALUES (1, 101, 'Single', 100.00);

INSERT INTO rooms (id, number, type, price) VALUES (2, 102, 'Double', 150.00);

INSERT INTO rooms (id, number, type, price) VALUES (3, 103, 'Suite', 200.00);


//server.js
require('dotenv').config();

const express = require('express');

const { Pool } = require('pg');

const bcrypt = require('bcrypt');

const jwt = require('jsonwebtoken');

const path = require('path');


const app = express();

const port = 3000;


const pool = new Pool({

  user: process.env.DB_USER,

  host: process.env.DB_HOST,

  database: process.env.DB_NAME,

  password: process.env.DB_PASSWORD,

  port: process.env.DB_PORT,

});


app.use(express.json());

app.use(express.static(path.join(__dirname, 'public')));


app.get('/', (req, res) => {

  res.sendFile(path.join(__dirname, 'public', 'index.html'));

});


pool.connect((err) => {

  if (err) {

    console.error('Error connecting to the database', err.stack);

  } else {

    console.log('Connected to the database');

  }

});


// Маршруты для добавления данных

app.post('/rooms', async (req, res) => {

  const { number, type, price } = req.body;

  const result = await pool.query(

    'INSERT INTO rooms (number, type, price) VALUES ($1, $2, $3) RETURNING *',

    [number, type, price]

  );

  res.json(result.rows[0]);

});


app.post('/bookings', async (req, res) => {

  const { room_id, guest_name, check_in_date, check_out_date } = req.body;

  const result = await pool.query(

    'INSERT INTO bookings (room_id, guest_name, check_in_date, check_out_date) VALUES ($1, $2, $3, $4) RETURNING *',

    [room_id, guest_name, check_in_date, check_out_date]

  );

  res.json(result.rows[0]);

});


// Маршруты для изменения данных

app.put('/rooms/:id', async (req, res) => {

  const { id } = req.params;

  const { number, type, price } = req.body;

  const result = await pool.query(

    'UPDATE rooms SET number = $1, type = $2, price = $3 WHERE id = $4 RETURNING *',

    [number, type, price, id]

  );

  res.json(result.rows[0]);

});


app.put('/bookings/:id', async (req, res) => {

  const { id } = req.params;

  const { room_id, guest_name, check_in_date, check_out_date } = req.body;

  const result = await pool.query(

    'UPDATE bookings SET room_id = $1, guest_name = $2, check_in_date = $3, check_out_date = $4 WHERE id = $5 RETURNING *',

    [room_id, guest_name, check_in_date, check_out_date, id]

  );

  res.json(result.rows[0]);

});


// Маршруты для удаления данных

app.delete('/rooms/:id', async (req, res) => {

  const { id } = req.params;

  await pool.query('DELETE FROM rooms WHERE id = $1', [id]);

  res.sendStatus(204);

});


app.delete('/bookings/:id', async (req, res) => {

  const { id } = req.params;

  await pool.query('DELETE FROM bookings WHERE id = $1', [id]);

  res.sendStatus(204);

});


// Маршруты для вывода данных

app.get('/rooms', async (req, res) => {

  const result = await pool.query('SELECT * FROM rooms');

  res.json(result.rows);

});


app.get('/bookings', async (req, res) => {

  const result = await pool.query('SELECT * FROM bookings');

  res.json(result.rows);

});


// Маршруты для регистрации и авторизации

app.post('/register', async (req, res) => {

  const { username, password } = req.body;

  const hashedPassword = await bcrypt.hash(password, 10);

  const result = await pool.query(

    'INSERT INTO users (username, password) VALUES ($1, $2) RETURNING *',

    [username, hashedPassword]

  );

  res.json(result.rows[0]);

});


app.post('/login', async (req, res) => {

  const { username, password } = req.body;

  const result = await pool.query('SELECT * FROM users WHERE username = $1', [username]);

  const user = result.rows[0];


  if (user && await bcrypt.compare(password, user.password)) {

    const token = jwt.sign({ id: user.id, username: user.username }, process.env.JWT_SECRET);

    res.json({ token });

  } else {

    res.status(401).send('Invalid credentials');

  }

});


// Middleware для проверки токена

const authenticateToken = (req, res, next) => {

  const token = req.headers['authorization'];

  if (!token) return res.sendStatus(401);


  jwt.verify(token, process.env.JWT_SECRET, (err, user) => {

    if (err) return res.sendStatus(403);

    req.user = user;

    next();

  });

};


// Пример защищенного маршрута

app.get('/protected', authenticateToken, (req, res) => {

  res.send('This is a protected route');

});


app.listen(port, () => {

  console.log(`Server is running on http://localhost:${port}`);

});



app.get('/rooms', async (req, res) => {

  try {

    const result = await pool.query('SELECT * FROM rooms');

    res.json(result.rows);

  } catch (err) {

    console.error(err);

    res.status(500).json({ error: 'Error fetching rooms' });

  }

});
